### PR TITLE
Create Fluid Force Computation class

### LINKED ISSUE #307: Create Fluid Force Computation class
### Purpose

To implement the Python classes that compute the lift and drag forces acting on a medium. See [this](https://ubcsailbot.atlassian.net/wiki/spaces/prjt22/pages/1809678337/Physics+Engine+Fluid+Simulation+Design+Specification#Fluid-Simulation) confluence page for more details.

### Description

You will be implementing the methods in a template class called `MediumForceComputation` located in `boat_simulator/nodes/physics_engine/fluid_forces.py`.

The class named `MediumForceComputation` has the following properties:
- Class Attributes:
	- `lift_coefficients` is a 2D array functioning as a lookup table for lift coefficients corresponding to various angles of attack.
	- `drag_coefficients` is a 2D array functioning as a lookup table for drag coefficients corresponding to various angles of attack.
	- `areas` is a 2D array functioning as a lookup table for the fluid-affected areas corresponding to different angles of attack.
	- `fluid_density` is the density of the fluid acting on the medium.
- Class Methods
    - `MediumForceComputation` is a constructor that accepts the lift coefficients, drag coefficients, affected areas, and fluid density as arguments.
    - `compute` is a function that accepts two arguments, `apparent_velocity`, which is the relative velocity between the fluid and the medium calculated by `fluid_velocity - medium_velocity`, and `orientation` which is the orientation angle of the medium.
    - `interpolate` is a function that accepts one argument, `attack_angle`, which is the angle of attack formed between orientation angle of the medium and the direction of the apparent velocity.

For reference, consider this diagram of an example medium (a sail) being acted on by an example fluid (wind):

Screenshot 2024-01-31 at 6 43 13 PM

Where
$\vec{V_w}$ is the velocity of the true wind acting on the medium.
$\vec{V_m}$ is the velocity of the medium.
$\vec{V_a}$ is the velocity of the [apparent](https://sailing-blog.nauticed.org/what-is-apparent-wind/) wind acting on the medium.
$\vec{F_L}$ is the lift force acting on the medium.
$\vec{F_D}$ is the drag force acting on the medium.

With the formulas
$F_L = f_L(\alpha) = \dfrac{1}{2}\rho ||v_A||^2 A(\alpha) C_L(\alpha)$, with direction perpendicular to the apparent velocity ($\vec{V_a}$)
$F_D = f_D(\alpha)$, with direction parallel to the apparent velocity ($\vec{V_a}$)

Where
$\rho$ is the fluid density.
$A$ is the area on the medium affected by the fluid, and is a function of $\alpha$.
$v_A$ is the apparent velocity between the fluid and the medium.
$C_L$ is the lift coefficient of the medium, and is a function of $\alpha$.
$\alpha$ is the angle of attack.

### Tests
Add your unit tests to `tests/unit/nodes/physics_engine/`. You will use the pytest API for testing, which is already installed.

Once complete, make a pull request to the main branch.

### Resources

- [Forces on a Sail](https://en.wikipedia.org/wiki/Forces_on_sails)
- [UBC Sailbot Python Conventions and Style Guide](https://ubcsailbot.github.io/docs/reference/python/conventions/)
- [Pytest Examples](https://docs.pytest.org/en/7.1.x/getting-started.html#get-started)

### PR DESCRIPTION
**Pull request by tanmaythakral**
### Description

- Resolves #307

https://github.com/UBCSailbot/boat_simulator/pull/76

### Verification

Screenshot 2024-03-21 180957
Screenshot 2024-03-21 181115
Screenshot 2024-03-21 181233
Screenshot 2024-03-21 181338
 

### Resources

- 
https://www1.grc.nasa.gov/beginners-guide-to-aeronautics/foilsimstudent/

### CODE REVIEW

**Comment by DFriend01** [comment]
Looks great! Can you add one more picture where the apparent wind lies in the first quadrant?

File: src/boat_simulator/tests/unit/nodes/physics_engine/test_fluid_forces.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        # Edge cases where orientation and velocity directions are opposite or identical
+        (np.array([1, 0]), 180, -180),
+        (np.array([-1, 0]), 180, 0),
+        (np.array([0, 1]), 270, -180),
+        (np.array([0, -1]), 90, -180),
+    ],
+)
+def test_calculate_attack_angle(
+    apparent_velocity, orientation, expected_angle, medium_force_setup
+):
```
Comment (DFriend01) [inline src/boat_simulator/tests/unit/nodes/physics_engine/test_fluid_forces.py:59]: Try test cases where the apparent velocity is not a unit vector

**Review by DFriend01 — CHANGES_REQUESTED** [review]

File: src/boat_simulator/boat_simulator/nodes/physics_engine/fluid_forces.py [hunk 2]
```diff
@@ -29,14 +30,48 @@ def __init__(
...
+        if np.all(apparent_velocity == 0):
+            # Directly return the normalized orientation as the angle of attack
+            # Normalize orientation to be within [-180, 180)
+            return ((orientation + 180) % 360) - 180
+
+        # Calculate the angle in degrees of the apparent velocity
+        angle_of_attack_raw = np.rad2deg(np.arctan2(apparent_velocity[1], apparent_velocity[0]))
+
+        # Adjust orientation to be in the range of [-180, 180)
+        orientation = ((orientation + 180) % 360) - 180
```
[the remark this answers is not in the document]
Comment (DFriend01) [reply src/boat_simulator/boat_simulator/nodes/physics_engine/fluid_forces.py:65]: @eomielan Apparently this function isn't working properly. Can you discuss with @tanmaythakral ?

**Review by DFriend01 — APPROVED** [review]
There are some failing jobs, but once you resolve those feel free to merge this PR. Great work!
